Add more ceph OSDs to support single node jobs

katarimanojk opened the pull request:
This patch add more block devices there by adding more OSDs in the ceph cluster to accomodate more ceph resources

As a pull request owner and reviewers, I checked that:
- [x] Appropriate testing is done and actually running
- [x] Appropriate documentation exists and/or is up-to-date:
  - [x] README in the role
  - [x] Content of the docs/source is reflecting the changes

katarimanojk commented on playbooks/ceph.yml:
```diff
@@ -87,6 +87,18 @@
+    - role: cifmw_block_device
+      cifmw_block_device_image_file: /var/lib/ceph-osd-1.img
+      cifmw_block_device_loop: /dev/loop4
+      cifmw_block_lv_name: ceph_lv_data1
```
Done

fultonj reviewed (commented):
Could you please generalize this? I've provided a quick playbook as an example.

In the playbook you'll want to replace calls to `debug` with calls to the role to create the block device and within the ceph spec role you'll need to introduce the jinja2 template.

Then please introduce a parameter to both the ceph playbook and the ceph_spec role for the desired number of OSDs per host. 

If `num_osds` is not provided then default it to 3 but only if there is only one ceph node. Else default it to 1. Those types of defaults should work for the new minimal 1 ceph node job but also not result in changes for the existing 3 node ceph jobs. 

```
---
- name: Test example
  gather_facts: false
  hosts: localhost ansible_connection=local
  vars:
    num_osds: 3
  tasks:
    - debug:
        msg: >-
          {{ cifmw_block_device_image_file }}
          {{ cifmw_block_device_loop }}
          {{ cifmw_block_lv_name }} 
          {{ cifmw_block_vg_name }}
          {{ cifmw_block_systemd_unit_file }}
      vars:
        cifmw_block_device_image_file: /var/lib/ceph-osd-{{ i }}.img
        cifmw_block_device_loop: /dev/loop{{ i + 3 }}
        cifmw_block_lv_name: ceph_lv{{ i }}
        cifmw_block_vg_name: ceph_vg{{ i }}
        cifmw_block_systemd_unit_file: /etc/systemd/system/ceph-osd-losetup-{{ i }}.service
      loop_control:
        loop_var: i
      loop: "{{ range(0, num_osds) }}"

    - debug:
        msg: >-
          {% for i in range(0, num_osds) %}
          - /dev/ceph_vg_data{{ i }}/ceph_lv_data{{ i }}
          {% endfor %}
```
When I run it I see this:
```
PLAY [Test example] **************************************************************************

TASK [debug] *********************************************************************************
ok: [localhost] => (item=0) => {
    "msg": "/var/lib/ceph-osd-0.img /dev/loop3 ceph_lv0  ceph_vg0 /etc/systemd/system/ceph-osd-losetup-0.service"
}
ok: [localhost] => (item=1) => {
    "msg": "/var/lib/ceph-osd-1.img /dev/loop4 ceph_lv1  ceph_vg1 /etc/systemd/system/ceph-osd-losetup-1.service"
}
ok: [localhost] => (item=2) => {
    "msg": "/var/lib/ceph-osd-2.img /dev/loop5 ceph_lv2  ceph_vg2 /etc/systemd/system/ceph-osd-losetup-2.service"
}

TASK [debug] *********************************************************************************
ok: [localhost] => {
    "msg": " - /dev/ceph_vg_data0/ceph_lv_data0  - /dev/ceph_vg_data1/ceph_lv_data1  - /dev/ceph_vg_data2/ceph_lv_data2 "
}

PLAY RECAP ***********************************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0 
```

fultonj commented on playbooks/ceph.yml:
```diff
@@ -87,6 +87,20 @@
   roles:
     - role: cifmw_block_device
+    # create more virtual block devices to have more OSDs so that
+    # we can accomodate more ceph resources and avaoid PG errors.
```
nit `avoid`

katarimanojk commented:
> Could you please generalize this? I've provided a quick playbook as an example.
> 
> In the playbook you'll want to replace calls to `debug` with calls to the role to create the block device and within the ceph spec role you'll need to introduce the jinja2 template.
> 
> Then please introduce a parameter to both the ceph playbook and the ceph_spec role for the desired number of OSDs per host.
> 
> If `num_osds` is not provided then default it to 3 but only if there is only one ceph node. Else default it to 1. Those types of defaults should work for the new minimal 1 ceph node job but also not result in changes for the existing 3 node ceph jobs.
> 
> ```
> ---
> - name: Test example
>   gather_facts: false
>   hosts: localhost ansible_connection=local
>   vars:
>     num_osds: 3
>   tasks:
>     - debug:
>         msg: >-
>           {{ cifmw_block_device_image_file }}
>           {{ cifmw_block_device_loop }}
>           {{ cifmw_block_lv_name }} 
>           {{ cifmw_block_vg_name }}
>           {{ cifmw_block_systemd_unit_file }}
>       vars:
>         cifmw_block_device_image_file: /var/lib/ceph-osd-{{ i }}.img
>         cifmw_block_device_loop: /dev/loop{{ i + 3 }}
>         cifmw_block_lv_name: ceph_lv{{ i }}
>         cifmw_block_vg_name: ceph_vg{{ i }}
>         cifmw_block_systemd_unit_file: /etc/systemd/system/ceph-osd-losetup-{{ i }}.service
>       loop_control:
>         loop_var: i
>       loop: "{{ range(0, num_osds) }}"
> 
>     - debug:
>         msg: >-
>           {% for i in range(0, num_osds) %}
>           - /dev/ceph_vg{{ i }}/ceph_lv{{ i }}
>           {% endfor %}
> ```
> 
> When I run it I see this:
> 
> ```
> PLAY [Test example] **************************************************************************
> 
> TASK [debug] *********************************************************************************
> ok: [localhost] => (item=0) => {
>     "msg": "/var/lib/ceph-osd-0.img /dev/loop3 ceph_lv0  ceph_vg0 /etc/systemd/system/ceph-osd-losetup-0.service"
> }
> ok: [localhost] => (item=1) => {
>     "msg": "/var/lib/ceph-osd-1.img /dev/loop4 ceph_lv1  ceph_vg1 /etc/systemd/system/ceph-osd-losetup-1.service"
> }
> ok: [localhost] => (item=2) => {
>     "msg": "/var/lib/ceph-osd-2.img /dev/loop5 ceph_lv2  ceph_vg2 /etc/systemd/system/ceph-osd-losetup-2.service"
> }
> 
> TASK [debug] *********************************************************************************
> ok: [localhost] => {
>     "msg": " - /dev/ceph_vg0/ceph_lv0  - /dev/ceph_vg1/ceph_lv1  - /dev/ceph_vg2/ceph_lv2 "
> }
> 
> PLAY RECAP ***********************************************************************************
> localhost                  : ok=2    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0 
> ```

I implemented this in the latest commit.


> Could you please generalize this? I've provided a quick playbook as an example.
> 
> In the playbook you'll want to replace calls to `debug` with calls to the role to create the block device and within the ceph spec role you'll need to introduce the jinja2 template.
> 
> Then please introduce a parameter to both the ceph playbook and the ceph_spec role for the desired number of OSDs per host.
> 
> If `num_osds` is not provided then default it to 3 but only if there is only one ceph node. Else default it to 1. Those types of defaults should work for the new minimal 1 ceph node job but also not result in changes for the existing 3 node ceph jobs.
> 
> ```
> ---
> - name: Test example
>   gather_facts: false
>   hosts: localhost ansible_connection=local
>   vars:
>     num_osds: 3
>   tasks:
>     - debug:
>         msg: >-
>           {{ cifmw_block_device_image_file }}
>           {{ cifmw_block_device_loop }}
>           {{ cifmw_block_lv_name }} 
>           {{ cifmw_block_vg_name }}
>           {{ cifmw_block_systemd_unit_file }}
>       vars:
>         cifmw_block_device_image_file: /var/lib/ceph-osd-{{ i }}.img
>         cifmw_block_device_loop: /dev/loop{{ i + 3 }}
>         cifmw_block_lv_name: ceph_lv{{ i }}
>         cifmw_block_vg_name: ceph_vg{{ i }}
>         cifmw_block_systemd_unit_file: /etc/systemd/system/ceph-osd-losetup-{{ i }}.service
>       loop_control:
>         loop_var: i
>       loop: "{{ range(0, num_osds) }}"
> 
>     - debug:
>         msg: >-
>           {% for i in range(0, num_osds) %}
>           - /dev/ceph_vg{{ i }}/ceph_lv{{ i }}
>           {% endfor %}
> ```
> 
> When I run it I see this:
> 
> ```
> PLAY [Test example] **************************************************************************
> 
> TASK [debug] *********************************************************************************
> ok: [localhost] => (item=0) => {
>     "msg": "/var/lib/ceph-osd-0.img /dev/loop3 ceph_lv0  ceph_vg0 /etc/systemd/system/ceph-osd-losetup-0.service"
> }
> ok: [localhost] => (item=1) => {
>     "msg": "/var/lib/ceph-osd-1.img /dev/loop4 ceph_lv1  ceph_vg1 /etc/systemd/system/ceph-osd-losetup-1.service"
> }
> ok: [localhost] => (item=2) => {
>     "msg": "/var/lib/ceph-osd-2.img /dev/loop5 ceph_lv2  ceph_vg2 /etc/systemd/system/ceph-osd-losetup-2.service"
> }
> 
> TASK [debug] *********************************************************************************
> ok: [localhost] => {
>     "msg": " - /dev/ceph_vg0/ceph_lv0  - /dev/ceph_vg1/ceph_lv1  - /dev/ceph_vg2/ceph_lv2 "
> }
> 
> PLAY RECAP ***********************************************************************************
> localhost                  : ok=2    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0 
> ```

i implemented in this approach in the latest commit.
cifmw_num_osds_perhost is generated based on the number of nodes and used while creating the block devices in ceph playbook and cifmw_ceph_spec.

There are few debug tasks that will be removed later.

fultonj reviewed (commented):
This looks good. Have you tested it on both 3 node and 1 node deployments?

katarimanojk commented:
recheck

cjeanner commented on playbooks/ceph.yml:
```diff
@@ -83,10 +83,35 @@
-  roles:
-    - role: cifmw_block_device
+  tasks:
+    - name: set cifmw_num_osds_perhost
```
In order to pass pre-commit tests, `All names should start with an uppercase letter`
This fails multiple times in the following files:
```
playbooks/ceph.yml:87: name[casing]: All names should start with an uppercase letter.
playbooks/ceph.yml:99: name[casing]: All names should start with an uppercase letter.
roles/cifmw_ceph_spec/tasks/main.yml:17: name[casing]: All names should start with an uppercase letter.
roles/cifmw_ceph_spec/tasks/main.yml:30: name[casing]: All names should start with an uppercase letter
```

Care to correct? It should help passing Prow jobs.

katarimanojk replied:
Sure, i will fix it in the next patch.

cjeanner commented on playbooks/ceph.yml:
```diff
@@ -83,10 +83,35 @@
+    # more ceph resources and avoid PG errors.
+      ansible.builtin.set_fact:
+        cifmw_num_osds_perhost: |
+          {% if groups[cifmw_ceph_target | default('computes')] | length == 1 %}
```
This fails in molecule tests - guess you'll have to fake the ansible inventory in order to inject a "computes" entry?

`The task includes an option with an undefined variable. The error was: 'dict object' has no attribute 'computes'. 'dict object' has no attribute 'computes'`

katarimanojk replied:
Thanks , i will fix in them in the next patch.

fultonj commented:
> This looks good. Have you tested it on both 3 node and 1 node deployments?

As per our conversation I undrstand this s

fultonj commented:
+2

fultonj commented on roles/cifmw_ceph_spec/templates/ceph_spec.yml.j2:
```diff
@@ -28,7 +28,11 @@ service_id: mgr
 service_name: mgr
 service_type: mgr
 ---
-{{ cifmw_ceph_spec_data_devices }}
```
@katarimanojk can you change this so that if the user provides `cifmw_ceph_spec_data_devices` that it uses it?

I think we should default the devices list the way you have it with a loop but that if a user provides the variable with specific block devices, then it should use those block devices.

katarimanojk replied:
@fultonj agree with you, latest patch will address it.
if `cifmw_ceph_spec_data_devices` is passed by the user, spec will be generated using it.
otherwise it defaults to the logic that uses the devices list in a loop based on `cifmw_num_osds_perhost`

morucci commented:
recheck

morucci commented:
recheck

fultonj reviewed (approved):
Thanks Manoj

cjeanner commented:
/approve
